keps/sig-cloud-provider/0002-cloud-controller-manager.md

```diff
@@ -266,9 +298,317 @@ In case of the cloud-controller-manager, the deployment should be deleted using
+
+We need a solution which meets the objective in both the short and long term. In the short term we cannot delete CCM or 
+cloud provider code from the K8s/K8s repo. We need to keep this code in K8s/K8s while we still support cloud provider 
+deployments from K8s/K8s. In the long term this code should be part of each cloud providers repo and that code should 
+be removed from K8s/K8s. This suggests that in the short term that code should have one source of truth. However it 
+should probably not end up in the vendor directory as that is not its intended final home. Other code such as the 
+controllers should end up in the vendor directory. Additionally each provider will need their own copy of 
+pkg/cloudprovider/providers/providers.go and related build file to properly control which Cloud Provider Implementation 
+get linked in.
+
+We also need to be able to package a combination of binaries from K8s/K8s and K8s/cloud-provider-<cp> into a deployable 
+package. The code for this will need to accommodate things like differing side cars for each cloud provider’s CSI 
+implementation and possible desire to run additional controller managers or extension api servers. As such it seems 
+better to have this code live in the cloud provider specific repo. This also allows this code to be simpler as it does 
+not have to attempt to support all the different cloud provider configurations. This separate from things like the 
+local deployment option, which K8s/K8s should continue to support.
+
+Lastly there are specific flags which need to be set on various binaries for this to work. Kubernetes API-Server, 
+Kubernetes Controller-Manager and Kubelet should all have the --cloud-provider flag set to external. For the Cloud 
+Controller-Manager the --cloud-provider flag should be set appropriately for that cloud provider. In addition we need 
+to set the set of controllers running in the Kubernetes Controller-Manager. More on that later.
+
+For further background material please look at [running cloud controller](https://kubernetes.io/docs/tasks/administer-cluster/running-cloud-controller/).
+
+### Design Options
+
+For the short term problem of sharing code between K8s/K8s repo, the K8s/cloud-provider repos and K8s SIGs/library repos; 
+there are 2 fundamental solutions. We can push more code into Staging to make the code available to the 
+K8s/cloud-provider repos. Currently for code which needs to be shared between K8s/K8s itself and other projects we put 
+the code in Staging. This allows the build machinery to detect things like attempts to make the shared code depend on 
+code which is not shared (i.e. disallowing code in Staging from depending Iton code in K8s/K8s but not in Staging) 
+In addition using Staging means that we can benefit from work to properly break out the libraries/code which exists in 
+Staging. Most of the Staging SIGs/library repos. The K8s/cloud-provider repos should be the top of any build dependency 
```

Comment:
We have to hash this out a bit more, but I _think_ what we want as an intermediate step is to move all the code in `k8s/k8s/pkg/cloudprovider/` to `k8s/k8s/staging/pkg/cloudprovider` which you then import from `k8s/cloud-provider-*`. We should avoid importing `k8s/cloud-provider-*` in `k8s/k8s`, but we may end up making exceptions while we transition to account for dependencies in `k8s/k8s` (though in theory, we should have `k8s/k8s/staging/cloudprovider` for as long as we have those dependencies).  

For other sig owned repos / external projects you would import `k8s/k8s/staging/pkg/cloudprovider` until we're ready to move everything into `k8s/cloud-provider-*`.

Reply:
+1 to Andrews explanation. 

Additonally I'm trying to set up a dependency ordering to avoid cyclic graphs and similar problems. So I'm suggesting we have a final state something like K8s/cloud-provider-aws (This is the repo which builds the deployable) which depends on K8s/K8s (the repo containing the K8s "kernel") and K8s Sigs/storage-csi-aws (the aws csi side car) which depends on K8s Sigs/cp-aws (the aws cloud provider interface). Right now I just want to be clear about ensuring that we do create any K8s/K8s  dependencies on K8s/cloud-provider-*.